3604/0/triangle.rb

```diff
@@ -0,0 +1,50 @@
+# Triangle Project Code.
+
+# Triangle analyzes the lengths of the sides of a triangle
+# (represented by a, b and c) and returns the type of triangle.
+#
+# It returns:
+#   :equilateral  if all sides are equal
+#   :isosceles    if exactly 2 sides are equal
+#   :scalene      if no sides are equal
+#
+# The tests for this method can be found in
+#   about_triangle_project.rb
+# and
+#   about_triangle_project_2.rb
+#
+
+class Triangle
```

Comment:
It looks pretty good 👍But we can make it even better.

1) First of all method `test_error`. Don't you think it should be private? We use it inside class only.  
2) It would be cool if we move all conditions to separate methods. Look what I mean:
```
def test_error
  raise TriangleError if @val_a <= 0 || @val_b <= 0 || @val_c <= 0
  sum = (@val_a + @val_b + @val_c) / 2.0
  raise TriangleError if ((sum - @val_a) * (sum - @val_b) * (sum - @val_c)) <= 0
end
```
When I read this code is hard for me because if want to understand what id do, I should understand whole logic. But you can make my life easer. 🙂Let's try it.
```
def test_error
  raise TriangleError if hasNegativeSides? || someConditionAgain?
end

private

def hasNegativeSides?
  @val_a <= 0 || @val_b <= 0 || @val_c <= 0
end

def someConditionAgain?
  ((sum - @val_a) * (sum - @val_b) * (sum - @val_c)) <= 0
end

def sum
  @sum ||= (@val_a + @val_b + @val_c) / 2.0
end
```
So what we got:
- now I shouldn't understand logic of this code. I see 'raise if it hasNegativeSides or someConditionAgain'. It's enough for me, I even shouldn't see implementation of it, this name helps me to read this code.
- we can use these conditions in whole class. For example if we will need `@val_a <= 0 || @val_b <= 0 || @val_c <= 0` again we will not duplicate code. We will just call `hasNegativeSides?`.

Please pay attention on this part:
```
def sum
  @sum ||= (@val_a + @val_b + @val_c) / 2.0
end
```
It's vital. 

What you do:
```
  sum = ((@val_a + @val_b + @val_c) / 2.0)
```
The thing is that you can use it in this scope only. (In `test_error` method). But what are you going to do, if you need it in another method too? Are you going to duplicate code? 🙂

So what I do: I move this logic to separate private method. And I can use it in class everywhere. But here new issue! In this code: `raise TriangleError if ((sum - @val_a) * (sum - @val_b) * (sum - @val_c)) <= 0` you need sum 3 times. So I think you understand issue: will we call our method 3 times? This means that the computer should calculate it 3 times, and earlier it calculated once! Here I can introduce you your new best friend - Memoization.

Look again
```
def sum
  @sum ||= (@val_a + @val_b + @val_c) / 2.0
end
```
What means this strange `@sum ||=`? 🤔It's easy. You tell to your method: "Okay, please calculate it if you haven't yet. But if you have already calculated it just use previous value" Cool? 🙂

Please do the same with method `triangle`. I'd like to see clear code there too.

Let me know if something is not clear for you.

Reply:
Thank you very much for the detailed answer.
I rewrote.

I wanted to do something similar:
`:equilateral if equilateral?
 :isosceles if isosceles?
 :scalene if scalene?`
but this code have error. I did differently.

Reply:
Let's not to change it 😉We can do this code better, but I think that enough for first review. Only one thing. For private scope: you shouldn't put `private` before each method.

It should looks like this:
```
class Something
  def method1 # this is public method
  end

  private

  def method1 # this is private method
  end

  def method1 # this is private method
  end
end
```

Just move all private methods under the `private` word

Reply:
And `some_condition_again?` is was just an example 😀Please give normal name for it

Reply:
@MargaretOm ^